### ISSUE TITLE
FEAT: Maintain multiple database connections

### ISSUE THREAD

**Issue by ReynoldsCahoon** [beekeeper-studio/beekeeper-studio#1667]
**Is your feature request related to a problem? Please describe.**
I frequently move between a few databases on the same connection. When I open up a tab to display a table, change databases and open a new tab for another table, when I swap back to the old tab I get the message, "This table does not exist in the selected database" error message, and a prompt to close the tab. It requires me to reselect the other database from the list, which then re-queries for the list of tables, and generally slows down my workflow.

**Describe the solution you'd like**
I would like an option to keep previous databases opened, when they're on the same connection, so that I can swap between queries/results without having to manually swap the database I'm looking at.

**Describe alternatives you've considered**
Other apps I've used will create a database 'tab' that keeps all queries and tabs made in each database separate. It'd also probably be sufficient to simply maintain the connection to any databases with associated tabs, and then swap the selected database in the list when you select a tab using a different database.

**Additional context**
Current experience.
<img width="1150" alt="image" src="https://github.com/beekeeper-studio/beekeeper-studio/assets/2556251/83537132-c43d-4bf5-b26e-8d54b45b5ec9">

TablePlus handles it by creating database tabs.
<img width="1112" alt="image" src="https://github.com/beekeeper-studio/beekeeper-studio/assets/2556251/378c9aee-de74-4214-802e-38f5fc243fd6">

**Database type, if appropriate**
Should work with all types, tested with MySQL.

**Comment by davidkaufman** [comment]
Hi @ReynoldsCahoon!

You can connect to a second database by pressing `Ctrl-Shift-N` (`Cmd-Shift-N` on Mac) to open a new Beekeeper Studio window.  You can also go to the File menu and select "New Window".

Many developers are familiar with this style of UI app design, as it's similar to many IDE's such as VS Code.  In Beekeeper Studio, each tab (whether a query editor, table view, schema view) is associated with the current window's current db connection, so having multiple tabs in one window that were connected to different databases could be confusing.

This is a frequently asked question (and frequently requested feature) so it's clear that many users don't know about it, and think that Beekeeper's unable to connect to multiple databases simultaneously.

As a long-time user I barely remember when i didn't know this, so I'm very interested in how to make this feature more visible so new users can more easily find it -- so I'm wondering:

- Where did you look when you were trying to find it?
- Where would you have expected to find a "connect to a new database in new window" feature?

Thanks in advance for your help making Beekeeper Studio better for everyone!

-dave

**Comment by ReynoldsCahoon** [comment]
Thanks, I wasn't aware you could open a second window.

I still maintain that this functionality should exist within the same window. The current solution is akin to asking the user to open a second instance of the application and create a new connection, whereas what I am requesting is that the application support this use case, as others do.

> _In Beekeeper Studio, each tab (whether a query editor, table view, schema view) is associated with the current window's current db connection, so having multiple tabs in one window that were connected to different databases could be confusing._

In case my original post was not clear, this is two databases on the same connection. I can write queries that join data from both of these databases currently on Beekeeper Studio (within a single window) and get results but I cannot view tables from each database without manually flipping between them.

Perhaps it was not intended, but your response reads as if the issue is with the user and not the software. My original request includes a UI solution found in a competing product, as well as the suggestion that _"It'd probably be sufficient to simply maintain the connection to any databases [that exist in the current connection] with associated tabs, and then swap the selected database in the list when you select a tab using a different database."_ I have inserted some additional context to help clarify what I'm suggesting.

If I can currently join two databases in Beekeeper Studio in a single query, why can I not also have a tab associated with one of those query-able databases that exists in the current connection display its values without having to manually change the selected database?

**Comment by rathboma** [comment]
Yeah I agree we need to support this somehow.

I don’t like the table plus approach. I think that sidebar should keep its use case.

Need to think about this some more.

**Comment by davidkaufman** [comment]
> many users don't know about it, and think that Beekeeper's unable to connect to multiple databases simultaneously

[low-hanging fruit] The simplest improvement(s) might be just to:
- change the text on the "New Window" menu (maybe only when already connected) to say "New database connection window", and/or
- add a link or button to the left nav when connected that says the same (or something similar) 

just so its a less hard-to-find feature...

**Comment by brunoantonieto** [comment]
hi there! i would like to give my two cents here, instead of adding stuff to the sidebar maybe we could just add a more intelligent downwards menu, to show other saved connections on top of `disconnect` and `edit`.
![image](https://github.com/beekeeper-studio/beekeeper-studio/assets/54121679/3a48bf24-4085-4d4c-833b-d42315561c76)

**Comment by rathboma** [comment]
@brunoantonieto You're a genius! I think we can just have a `Switch Connection` menu item. I think that'd be an easy thing to do before we work on the full thing

**Comment by rathboma** [comment]
Here's a strawman on how multi-connections might work (this is just a quick mockup). Thoughts welcome:

![Multiple Connections](https://github.com/user-attachments/assets/5cc1d6b3-6526-4768-b1b1-d001996ae221)

**Comment by marcosfede** [comment]
I would absolutely love this, I think this is my biggest pain point using the product (which I love). I manage multiple dbs and having so many windows open is not that manageable. I'd love to have this kind of "slackish" db switcher or any kind of tabs should suffice

**Comment by ahhandoyo** [comment]
> Hi [@ReynoldsCahoon](https://github.com/ReynoldsCahoon)!
> 
> You can connect to a second database by pressing `Ctrl-Shift-N` (`Cmd-Shift-N` on Mac) to open a new Beekeeper Studio window. You can also go to the File menu and select "New Window".
> 
> Many developers are familiar with this style of UI app design, as it's similar to many IDE's such as VS Code. In Beekeeper Studio, each tab (whether a query editor, table view, schema view) is associated with the current window's current db connection, so having multiple tabs in one window that were connected to different databases could be confusing.
> 
> This is a frequently asked question (and frequently requested feature) so it's clear that many users don't know about it, and think that Beekeeper's unable to connect to multiple databases simultaneously.
> 
> As a long-time user I barely remember when i didn't know this, so I'm very interested in how to make this feature more visible so new users can more easily find it -- so I'm wondering:
> 
> * Where did you look when you were trying to find it?
> * Where would you have expected to find a "connect to a new database in new window" feature?
> 
> Thanks in advance for your help making Beekeeper Studio better for everyone!
> 
> -dave

but it will take up quite a lot of memory if more and more new windows are opened. I've tried doing it on mac, and every time I open a new window my memory activity increases. and that doesn't seem too good.

**Comment by filereadAndrew** [comment]
is there any updates to do this? I would also like to see this as a feature

**Comment by danivpv** [comment]
Just adding my two cents since it is also my biggest pain point alongside having a keybinding to focus on the editor pane for keyboard navigation. I mention the latter since the experience I would like to enable is to be able to navigate only through keybindings and the most common workflow in my case is switching connections, tables, etc.. and then focusing in the table to edit or detail a specific field. 

I use MacOS and the multiple windows doesnt work well for keyboard navigation since cmd+tab will only show one window never both. I swear I would have already bought the subscription if it werent for this issues.

**Comment by davidkaufman** [comment]
@danivpv wrote:
> I use MacOS and the multiple windows doesnt work well for keyboard navigation since cmd+tab will only show one window never both. I swear I would have already bought the subscription if it werent for this issues.

If you're looking for a way to navigate between multiple open Beekeeper Studio windows without taking your hands off the keyboard, try CMD-tilde (on macOS -- not sure about other OS's that lack a CMD key tho... 😜). Works on most other macOS apps too!

_Edit: Or maybe it's CMD-backtick? I pronounce it CMD-tilde in my roof-brain-chatter ...as I do it ...which I do many, many times a day. But, for avoidance of doubt, and since it works to cycle windows backwards if you add the Shift key, it's these two:_  

<img width="760" height="420" alt="Image" src="https://github.com/user-attachments/assets/e9b8ebc0-a2ce-4b31-aca9-0ba0fb0a5730" />

**Comment by danivpv** [comment]
Yes thank you @davidkaufman . Really does help navigating! Wont take more of your time @everyone so.. Have a great day!

**Comment by zevada** [comment]
+1

This is one of my main pain points with Beekeeper as well. Adding something like a Slack workspace switcher would be a great addition to this product. I normally deal with 3 databases at the same time and having to rely on multiple windows feels clumsy.

**Comment by schmitty2535** [comment]
I agree, this feature needs to be added before I would purchase this product. I think its a major intentional limitation of the program. I'm coming from mysql workbench and Datagrip which both have the ability to query and display data from multiple schemas in one instance of the program without switching workspaces.

**Comment by calinortan** [comment]
just bumping this.

I see there is a switch option but it fails to connect

<img width="601" height="181" alt="Image" src="https://github.com/user-attachments/assets/8c745ee6-c902-4be9-9388-84c09286af17" />

**Comment by martijnwijnandsMES** [comment]
In SSMS you can just use [database1].[table1] and [database2].[tablex] in one query window. It would be handy to have this as well. Now it is basically unusable for me.
